### PR TITLE
Fix issue #55

### LINKED ISSUE #55: A fresh logfile is started even if the node is about to quit due to an already running instance
**Scenario**:
I run the attachment demo. 
All nodes run in background. 
I _cd_ to a node (e.g. nodea) directory and run _java -jar corda.jar_. 
Corda kindly checks if there is other application running from the same directory and exits. 

 **Issue** 
it starts to write to logs before making a check, what means that original logs are overwrite. I think they are properly archived into archive subdirectory (BTW I expected to find them in logs directory), but the ‘main’ logs file is gone.

### PR DESCRIPTION
**Pull request by exFalso**
This PR aims to fix #55

It factors out the file locking into a separate class `EnforceSingleNodeIsRunning`, which is started before the logging is initialised.

I added an integ test that checks that only a single node is writing to the logfile, though it's a bit brittle, it greps for "Logs can be found in". Thoughts?

Note also that `EnforceSingleNodeIsRunning` itself does some logging - as the logging is not initialised by this time it will use the default slf4j settings. We could leave it like this or do some manual printing instead

### CODE REVIEW

**Comment by exFalso** [comment]
Ah also did some cleanup on the way, on a double-node start the second node failure managed to get into a Driver codepath that made it skip some shutdown hooks, this is fixed now

File: node/src/integration-test/kotlin/net/corda/node/BootTests.kt [hunk 1]
```diff
@@ -23,6 +29,23 @@ class BootTests {
...
+        driver(isDebug = true, systemProperties = mapOf("log4j.configurationFile" to logConfigFile.toString())) {
+            val alice = startNode(ALICE.name).get()
+            val logFolder = alice.configuration.baseDirectory / "logs"
+            val logFile = logFolder.toFile().listFiles { _, name -> name.endsWith(".log") }.single()
+            // Start second Alice, should fail
+            assertThatThrownBy {
+                startNode(ALICE.name).getOrThrow()
+            }
+            // We count the number of nodes that wrote into the logfile by counting "Logs can be found in"
+            val numberOfNodesThatLogged = Files.lines(logFile.toPath()).filter { it.contains("Logs can be found in") }.count()
```
Comment (andr3ej) [inline node/src/integration-test/kotlin/net/corda/node/BootTests.kt]: Can we have a constant for the "Logs can be found in" string? It now exists in 3 places in the codebase.

File: node/src/main/kotlin/net/corda/node/Corda.kt [hunk 2]
```diff
@@ -65,6 +64,11 @@ fun main(args: Array<String>) {
         exitProcess(1)
     }
 
+    val enforceSingleNodeIsRunning = EnforceSingleNodeIsRunning(cmdlineOptions.baseDirectory)
+    // We do the single node check before we initialise logging so that in case of a double-node start it doesn't mess
+    // with the running node's logs.
+    enforceSingleNodeIsRunning.start()
```
Comment (andr3ej) [inline node/src/main/kotlin/net/corda/node/Corda.kt]: I'm not keen on the name "start", which suggests something is being backgrounded (the shutdown hook doesn't count IMO).

File: node/src/main/kotlin/net/corda/node/driver/Driver.kt [hunk 3]
```diff
@@ -331,20 +331,21 @@ class ShutdownManager(private val executorService: ExecutorService) {
...
-        shutdowns.reversed().forEach { shutdown ->
+        val shutdowns = shutdownFutures.map { ErrorOr.catch { it.get(1, SECONDS) } }
+        shutdowns.reversed().forEach { errorOrShutdown ->
             try {
-                shutdown()
+                if (errorOrShutdown.error == null) {
+                    errorOrShutdown.value?.invoke()
+                } else {
+                    log.error("Exception while getting shutdown method, disregarding", errorOrShutdown.error)
+                }
```
Comment (andr3ej) [inline node/src/main/kotlin/net/corda/node/driver/Driver.kt]: ErrorOr.match can do the logic here, and then you could pass in { shutdown -> shutdown() }

File: node/src/main/kotlin/net/corda/node/internal/EnforceSingleNodeIsRunning.kt [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+        val log = loggerFor<EnforceSingleNodeIsRunning>()
+    }
+
+    fun start() {
+        // Write out our process ID (which may or may not resemble a UNIX process id - to us it's just a string) to a
+        // file that we'll do our best to delete on exit. But if we don't, it'll be overwritten next time. If it already
+        // exists, we try to take the file lock first before replacing it and if that fails it means we're being started
+        // twice with the same directory: that's a user error and we should bail out.
+        val pidPath = baseDirectory / "process-id"
+        val pidFile = pidPath.toFile()
```
Comment (andr3ej) [inline node/src/main/kotlin/net/corda/node/internal/EnforceSingleNodeIsRunning.kt]: pidPath can be inlined, currently it looks like something else is going to use it.

File: node/src/main/kotlin/net/corda/node/internal/EnforceSingleNodeIsRunning.kt [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+        if (!pidFile.exists()) {
+            pidFile.createNewFile()
+        }
```
Comment (andr3ej) [inline node/src/main/kotlin/net/corda/node/internal/EnforceSingleNodeIsRunning.kt]: Logically this is a race, and I believe createNewFile will simply return false if we don't bother with the existence check.

File: node/src/main/kotlin/net/corda/node/Corda.kt (hunk 2 above)
Comment (exFalso) [inline node/src/main/kotlin/net/corda/node/Corda.kt]: Changed it to be simply a function

**Comment by exFalso** [comment]
Addressed comment back to you @andr3ej

**Comment by andr3ej** [comment]
Looking good, I'll approve after I've seen what it will take to make the build succeed.

**Comment by exFalso** [comment]
grr been running it in a loop can't repro

**Comment by exFalso** [comment]
ok reran it and now it's green, doesn't really inspire confidence, but i think it may have run it with the previous version somehow

**Pull request merged**